source/game/system/map/MapdataCheckPoint.hh

```diff
@@ -20,27 +30,55 @@ public:
 
     MapdataCheckPoint(const SData *data);
     void read(EGG::Stream &stream);
+    void initCheckpointLinks(MapdataCheckPointAccessor &accessor, int id);
+
+    /// @beginGetters
+    [[nodiscard]] s8 checkArea() const;
+    /// @endGetters
+
+    static constexpr s8 NORMAL_CHECKPOINT = -1; ///< Only used for picking respawn position
+    static constexpr s8 FINISH_LINE = 0;        ///< Triggers a lap change
 
 private:
+    [[nodiscard]] MapdataCheckPoint *nextPoint(size_t i) const;
+
     const SData *m_rawData;
     EGG::Vector2f m_left;
     EGG::Vector2f m_right;
-    u8 m_jugemIndex; ///< Index of respawn point associated with this checkpoint. Players who die
+    s8 m_jugemIndex; ///< Index of respawn point associated with this checkpoint. Players who die
                      ///< here will be respawned at this point.
+    /// Either:
+    /// - a @ref `NORMAL_CHECKPOINT` (0) used to calulate respawns,
+    /// - a @ref `FINISH_LINE` (-1) which updates the lap count when crossed, or
+    /// - a "key checkpoint" (1-127) used to ensure racers travel around the entire
+    /// course before proceeding to the next lap. the type value represents the index,
+    /// i.e. racers must pass checkpoint with @ref `m_type` 1, then 2, then 3 etc..
     s8 m_checkArea;
-    u8 m_prevPt;
-    u8 m_nextPt;
+    s8 m_prevPt;
+    s8 m_nextPt;
     u16 m_nextCount;
     u16 m_prevCount;
     EGG::Vector2f m_midpoint;
     EGG::Vector2f m_dir;
+    u16 m_id;
+    std::array<MapdataCheckPoint *, 6> m_prevPoints;
+    std::array<LinkedCheckpoint, 6> m_nextPoints;
```

Comment:
Still don't like that we have these hard-coded. Can we have a `static constexpr size_t` for these sizes?